UefiTestingPkg/Library/DriverDependencyLogging/DriverDependencyLogging.c

```diff
@@ -0,0 +1,423 @@
+/** @file -- DriverDependencyLogging.c
+
+This library and toolset are used with the Core DXE dispatcher to log all DXE drivers' protocol usage and
+dependency expression implementation during boot.
+
+See the readme.md file for full information
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include <PiDxe.h>
+
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/DebugLib.h>
+#include <Library/PrintLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiLib.h>
+#include "DriverDependencyLogging.h"
+
+//
+// WARNING - This library is tightly coupled to the MdeModulePkg Core DXE driver
+//
+#include "../Core/Dxe/DxeMain.h"

```

Comment:
Ron,

We need to talk about this.  
Relative includes especially back to DxeMain are not something we should do.   This opens up huge potential for problems.

Reply:
I agree this is not ideal, but since the code is heavily tied to the DxeMain and is intended to extend the functionality of DxeMain only, this was the best option.

Modifying the DxeMain code to expose a private structure is off the table for an option.  And making a copy of the structure definition locally can cause worse problems since the structure has a LIST_ENTRY embedded and if offsets change, we will have all kinds of random errors when typcasting from the linked list pointer to the data pointer.

Your question does bring up a good point though, maybe it is not clear how tightly this is coupled to the DxeMain driver.  I think I do need to maybe name it something like DxeMainDepexLoggingLib that shows it is a MU_PLUS item that is extending the DxeMain driver.